flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -298,6 +303,27 @@ private JobGraph createJobGraph() {
         return jobGraph;
     }
 
+    private void waitForSerializationFuturesAndUpdateJobVertices()
+            throws ExecutionException, InterruptedException {
+        for (JobVertex jobVertex : jobGraph.getVertices()) {
+            final List<CompletableFuture<SerializedValue<OperatorCoordinator.Provider>>>
+                    futuresForJobVertex = coordinatorSerializationFutures.remove(jobVertex.getID());
+            if (futuresForJobVertex == null) {
+                LOG.warn(
+                        "No OperatorCoordinator creator serialized for JobVertex {}.",
+                        jobVertex.getID());
+            } else {
+                FutureUtils.combineAll(futuresForJobVertex)
+                        .get()
+                        .forEach(jobVertex::addOperatorCoordinator);
+            }
+        }
+        Preconditions.checkState(
+                coordinatorSerializationFutures.isEmpty(),
+                "There are still serialization futures not processed for JobVertex instances: {}",
```

Comment:
This seems misleading. if this happens then some concurrent operation is going on.

Reply:
I rephrased the condition that should never occur... Were you suggesting that we should still wait for the concurrent operations to be completed? :thinking: I didn't implement that because this specific state would mean that there's some bug somewhere else. The Precondition should trigger the `IllegalStateException` causing the cluster to fail fatally.

Reply:
We shouldn't wait. If this happens something added a coordinator while we were already inserting them; and we should fail hard when that happens.